File: src/main/java/com/surrealdb/Geometry.java
package com.surrealdb;

import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.List;

/**
 * A SurrealDB geometry value. Supports all GeoJSON-style types: {@code Point},
 * {@code LineString}, {@code Polygon}, {@code MultiPoint}, {@code MultiLineString},
 * {@code MultiPolygon}, and {@code GeometryCollection}.
 * <p>
 * Coordinates use {@link java.awt.geom.Point2D.Double} where {@code x} is the
 * longitude and {@code y} is the latitude (GeoJSON ordering). Multi-coordinate
 * types are exposed as nested {@link List}s of points.
 * <p>
 * Instances can be read from query results (via {@link Value#getGeometry()}) and
 * constructed with the static factory methods to be sent back to the database.
 */
public class Geometry extends Native {

	Geometry(long ptr) {
		super(ptr);
	}

	// ---- native: read ----

	private static native boolean isPoint(long ptr);

	private static native double[] getPoint(long ptr);

	private static native String getType(long ptr);

	private static native long getCoordinates(long ptr);

	private static native long[] getCollection(long ptr);

	// ---- native: write ----

	private static native long newPoint(double x, double y);

	private static native long newLineString(double[] coords);

	private static native long newPolygon(double[][] rings);

	private static native long newMultiPoint(double[] coords);

	private static native long newMultiLineString(double[][] lines);

	private static native long newMultiPolygon(double[][][] polygons);

	private static native long newCollection(long[] geometryPtrs);

	@Override
	final native String toString(long ptr);

	@Override
	final native int hashCode(long ptr);

	@Override
	final native boolean equals(long ptr1, long ptr2);

	@Override
	final native void deleteInstance(long ptr);

	// ---- type discrimination ----

	/**
	 * Returns the geometry type: one of {@code "Point"}, {@code "LineString"},
	 * {@code "Polygon"}, {@code "MultiPoint"}, {@code "MultiLineString"},
	 * {@code "MultiPolygon"}, or {@code "GeometryCollection"}.
	 */
	public String getType() {
		return getType(getPtr());
	}

	public boolean isPoint() {
		return isPoint(getPtr());
	}

	public boolean isLineString() {
		return "LineString".equals(getType());
	}

	public boolean isPolygon() {
		return "Polygon".equals(getType());
	}

	public boolean isMultiPoint() {
		return "MultiPoint".equals(getType());
	}

	public boolean isMultiLineString() {
		return "MultiLineString".equals(getType());
	}

	public boolean isMultiPolygon() {
		return "MultiPolygon".equals(getType());
	}

	public boolean isGeometryCollection() {
		return "GeometryCollection".equals(getType());
	}

	// ---- readers ----

	public Point2D.Double getPoint() {
		final double[] coord = getPoint(getPtr());
		return new Point2D.Double(coord[0], coord[1]);
	}

	public List<Point2D.Double> getLineString() {
		return parseLine(coordinates());
	}

	public List<Point2D.Double> getMultiPoint() {
		return parseLine(coordinates());
	}

	/**
	 * Returns the polygon's rings. The first ring is the exterior boundary; any
	 * subsequent rings are interior holes.
	 */
	public List<List<Point2D.Double>> getPolygon() {
		return parseRings(coordinates());
	}

	public List<List<Point2D.Double>> getMultiLineString() {
		return parseRings(coordinates());
	}

	public List<List<List<Point2D.Double>>> getMultiPolygon() {
		final Array polygons = coordinates();
		final List<List<List<Point2D.Double>>> result = new ArrayList<>(polygons.len());
		for (int i = 0; i < polygons.len(); i++) {
			result.add(parseRings(polygons.get(i).getArray()));
		}
		return result;
	}

	public List<Geometry> getGeometryCollection() {
		final long[] ptrs = getCollection(getPtr());
		final List<Geometry> result = new ArrayList<>(ptrs.length);
		for (final long ptr : ptrs) {
			result.add(new Geometry(ptr));
		}
		return result;
	}

	// ---- factories (x = longitude, y = latitude) ----

	public static Geometry point(double x, double y) {
		return new Geometry(newPoint(x, y));
	}

	public static Geometry point(Point2D.Double point) {
		return new Geometry(newPoint(point.x, point.y));
	}

	public static Geometry lineString(List<Point2D.Double> points) {
		return new Geometry(newLineString(flatten(points)));
	}

	public static Geometry multiPoint(List<Point2D.Double> points) {
		return new Geometry(newMultiPoint(flatten(points)));
	}

	/**
	 * Builds a polygon. The first ring is the exterior boundary; any subsequent
	 * rings are interior holes.
	 */
	public static Geometry polygon(List<List<Point2D.Double>> rings) {
		return new Geometry(newPolygon(flattenRings(rings)));
	}

	public static Geometry multiLineString(List<List<Point2D.Double>> lines) {
		return new Geometry(newMultiLineString(flattenRings(lines)));
	}

	public static Geometry multiPolygon(List<List<List<Point2D.Double>>> polygons) {
		final double[][][] data = new double[polygons.size()][][];
		for (int i = 0; i < polygons.size(); i++) {
			data[i] = flattenRings(polygons.get(i));
		}
		return new Geometry(newMultiPolygon(data));
	}

	public static Geometry geometryCollection(List<Geometry> geometries) {
		final long[] ptrs = new long[geometries.size()];
		for (int i = 0; i < geometries.size(); i++) {
			ptrs[i] = geometries.get(i).getPtr();
		}
		return new Geometry(newCollection(ptrs));
	}

	// ---- helpers ----

	/** The geometry's coordinates as a nested array of {@code [x, y]} pairs. */
	private Array coordinates() {
		return new Value(getCoordinates(getPtr())).getArray();
	}

	private static Point2D.Double parsePoint(Value coordinate) {
		final Array xy = coordinate.getArray();
		return new Point2D.Double(xy.get(0).getDouble(), xy.get(1).getDouble());
	}

	private static List<Point2D.Double> parseLine(Array line) {
		final List<Point2D.Double> points = new ArrayList<>(line.len());
		for (int i = 0; i < line.len(); i++) {
			points.add(parsePoint(line.get(i)));
		}
		return points;
	}

	private static List<List<Point2D.Double>> parseRings(Array rings) {
		final List<List<Point2D.Double>> result = new ArrayList<>(rings.len());
		for (int i = 0; i < rings.len(); i++) {
			result.add(parseLine(rings.get(i).getArray()));
		}
		return result;
	}

	private static double[] flatten(List<Point2D.Double> points) {
		final double[] coords = new double[points.size() * 2];
		for (int i = 0; i < points.size(); i++) {
			coords[i * 2] = points.get(i).x;
			coords[i * 2 + 1] = points.get(i).y;
		}
		return coords;
	}

	private static double[][] flattenRings(List<List<Point2D.Double>> rings) {
		final double[][] data = new double[rings.size()][];
		for (int i = 0; i < rings.size(); i++) {
			data[i] = flatten(rings.get(i));
		}
		return data;
	}
}

File: src/main/rust/geometry.rs
use std::hash::{DefaultHasher, Hash, Hasher};
use std::ptr::null_mut;
use std::sync::Arc;

use crate::with_env_body;
use geo::{Coord, LineString, MultiLineString, MultiPoint, MultiPolygon, Point, Polygon};
use jni::objects::{JClass, JDoubleArray, JLongArray, JObjectArray};
use jni::sys::{jboolean, jdouble, jdoubleArray, jint, jlong, jlongArray, jstring};
use jni::{Env, EnvUnowned};
use surrealdb::types::{Geometry, Value};

use crate::error::SurrealError;
use crate::{
    get_long_array, get_value_instance, new_double_point, new_jlong_array, new_string,
    release_instance, JniTypes,
};

/// Unwraps a `Result` inside a `jlong`-returning native body, surfacing any error
/// (either a [`SurrealError`] or a `jni::errors::Error`) as a Java exception and
/// returning `0`. `SurrealError::from` is an identity conversion for `SurrealError`
/// and converts JNI errors, so this handles both error kinds.
macro_rules! unwrap_or_throw {
    ($env:expr, $result:expr) => {
        match $result {
            Ok(v) => v,
            Err(e) => return SurrealError::from(e).exception($env, || 0),
        }
    };
}

/// Reads a flat `double[]` of `[x0, y0, x1, y1, ...]` into a `Vec<Coord<f64>>`.
fn read_coords(env: &mut Env, arr: JDoubleArray) -> Result<Vec<Coord<f64>>, SurrealError> {
    let len = arr.len(env)?;
    let mut buf = vec![0.0_f64; len];
    arr.get_region(env, 0, &mut buf)?;
    let mut coords = Vec::with_capacity(len / 2);
    let mut i = 0;
    while i + 1 < len {
        coords.push(Coord {
            x: buf[i],
            y: buf[i + 1],
        });
        i += 2;
    }
    Ok(coords)
}

/// Reads a `double[][]` (each inner array a flat coordinate list) into a
/// `Vec<LineString<f64>>` — used for polygon rings and multi-line components.
fn read_rings(
    env: &mut Env,
    arr: JObjectArray<JDoubleArray>,
) -> Result<Vec<LineString<f64>>, SurrealError> {
    let len = arr.len(env)?;
    let mut rings = Vec::with_capacity(len);
    for i in 0..len {
        let inner = arr.get_element(env, i)?;
        rings.push(LineString::new(read_coords(env, inner)?));
    }
    Ok(rings)
}

/// Builds a polygon from its rings: the first ring is the exterior boundary and
/// any remaining rings are interior holes (matching GeoJSON / `as_coordinates`).
fn rings_to_polygon(mut rings: Vec<LineString<f64>>) -> Polygon<f64> {
    if rings.is_empty() {
        Polygon::new(LineString::new(vec![]), vec![])
    } else {
        let exterior = rings.remove(0);
        Polygon::new(exterior, rings)
    }
}

// ---------------------------------------------------------------------------
// Read side: type discrimination + coordinate extraction
// ---------------------------------------------------------------------------

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_deleteInstance<'local>(
    _env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jboolean {
    release_instance::<Arc<Value>>(ptr);
    true as jboolean
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_isPoint<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jboolean {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, || false as jboolean);
        if let Value::Geometry(g) = value.as_ref() {
            g.is_point() as jboolean
        } else {
            SurrealError::NullPointerException("Geometry").exception(env, || false as jboolean)
        }
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_getType<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jstring {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, null_mut);
        if let Value::Geometry(g) = value.as_ref() {
            return new_string!(env, g.as_type(), null_mut);
        }
        SurrealError::NullPointerException("Geometry").exception(env, null_mut)
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_getPoint<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jdoubleArray {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, null_mut);
        if let Value::Geometry(Geometry::Point(pt)) = value.as_ref() {
            return new_double_point!(env, pt, null_mut);
        }
        SurrealError::NullPointerException("Geometry").exception(env, null_mut)
    })
}

/// Returns the geometry's coordinates as a nested SurrealDB `Value` (arrays of
/// `[x, y]` doubles), mirroring GeoJSON nesting. The Java side traverses this with
/// the existing `Value`/`Array` API. Not used for `GeometryCollection`, whose
/// per-child types must be preserved — see `getCollection`.
#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_getCoordinates<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jlong {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, || 0);
        if let Value::Geometry(g) = value.as_ref() {
            return JniTypes::new_value(Arc::new(g.as_coordinates()));
        }
        SurrealError::NullPointerException("Geometry").exception(env, || 0)
    })
}

/// Returns the children of a `GeometryCollection` as an array of `Value` pointers,
/// each a `Value::Geometry`, so the Java side can wrap them as `Geometry` and keep
/// each child's own type.
#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_getCollection<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jlongArray {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, null_mut);
        if let Value::Geometry(Geometry::Collection(geoms)) = value.as_ref() {
            let ptrs: Vec<jlong> = geoms
                .iter()
                .map(|g| JniTypes::new_value(Arc::new(Value::Geometry(g.clone()))))
                .collect();
            return new_jlong_array!(env, &ptrs, null_mut);
        }
        SurrealError::NullPointerException("GeometryCollection").exception(env, null_mut)
    })
}

// ---------------------------------------------------------------------------
// Write side: construct a Value::Geometry from Java coordinates. Each native
// returns a pointer to an `Arc<Value>` (a `Value::Geometry`), identical to the
// read-side representation, so a single `Geometry` Java class serves both and
// `deleteInstance` releases it the same way.
// ---------------------------------------------------------------------------

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newPoint<'local>(
    _env: EnvUnowned<'local>,
    _class: JClass<'local>,
    x: jdouble,
    y: jdouble,
) -> jlong {
    JniTypes::new_value(Arc::new(Value::Geometry(Geometry::Point(Point::new(x, y)))))
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newLineString<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    coords: JDoubleArray<'local>,
) -> jlong {
    with_env_body!(env, env, {
        let coords = unwrap_or_throw!(env, read_coords(env, coords));
        JniTypes::new_value(Arc::new(Value::Geometry(Geometry::Line(LineString::new(
            coords,
        )))))
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newMultiPoint<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    coords: JDoubleArray<'local>,
) -> jlong {
    with_env_body!(env, env, {
        let coords = unwrap_or_throw!(env, read_coords(env, coords));
        let points: Vec<Point<f64>> = coords.into_iter().map(Point::from).collect();
        JniTypes::new_value(Arc::new(Value::Geometry(Geometry::MultiPoint(
            MultiPoint::new(points),
        ))))
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newPolygon<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    rings: JObjectArray<'local, JDoubleArray<'local>>,
) -> jlong {
    with_env_body!(env, env, {
        let rings = unwrap_or_throw!(env, read_rings(env, rings));
        JniTypes::new_value(Arc::new(Value::Geometry(Geometry::Polygon(
            rings_to_polygon(rings),
        ))))
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newMultiLineString<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    lines: JObjectArray<'local, JDoubleArray<'local>>,
) -> jlong {
    with_env_body!(env, env, {
        let lines = unwrap_or_throw!(env, read_rings(env, lines));
        JniTypes::new_value(Arc::new(Value::Geometry(Geometry::MultiLine(
            MultiLineString::new(lines),
        ))))
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newMultiPolygon<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    polygons: JObjectArray<'local, JObjectArray<'local, JDoubleArray<'local>>>,
) -> jlong {
    with_env_body!(env, env, {
        let len = unwrap_or_throw!(env, polygons.len(env));
        let mut polys = Vec::with_capacity(len);
        for i in 0..len {
            let rings_arr = unwrap_or_throw!(env, polygons.get_element(env, i));
            let rings = unwrap_or_throw!(env, read_rings(env, rings_arr));
            polys.push(rings_to_polygon(rings));
        }
        JniTypes::new_value(Arc::new(Value::Geometry(Geometry::MultiPolygon(
            MultiPolygon::new(polys),
        ))))
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_newCollection<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptrs: JLongArray<'local>,
) -> jlong {
    with_env_body!(env, env, {
        let ptrs = get_long_array!(env, &ptrs, || 0);
        let mut geoms = Vec::with_capacity(ptrs.len());
        for p in ptrs {
            let value = get_value_instance!(env, p, || 0);
            if let Value::Geometry(g) = value.as_ref() {
                geoms.push(g.clone());
            } else {
                return SurrealError::NullPointerException("Geometry").exception(env, || 0);
            }
        }
        JniTypes::new_value(Arc::new(Value::Geometry(Geometry::Collection(geoms))))
    })
}

// ---------------------------------------------------------------------------
// Object identity (equals / hashCode / toString) — delegated to the SDK's
// `Geometry` impls, which already cover every variant.
// ---------------------------------------------------------------------------

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_equals<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr1: jlong,
    ptr2: jlong,
) -> jboolean {
    with_env_body!(env, env, {
        let v1 = get_value_instance!(env, ptr1, || false as jboolean);
        let v2 = get_value_instance!(env, ptr2, || false as jboolean);
        if let (Value::Geometry(g1), Value::Geometry(g2)) = (v1.as_ref(), v2.as_ref()) {
            return g1.eq(g2) as jboolean;
        }
        SurrealError::NullPointerException("Geometry").exception(env, || false as jboolean)
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_hashCode<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jint {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, || 0);
        if let Value::Geometry(g) = value.as_ref() {
            let mut hasher = DefaultHasher::new();
            g.hash(&mut hasher);
            let hash64 = hasher.finish();
            return (hash64 & 0xFFFFFFFF) as jint;
        }
        SurrealError::NullPointerException("Geometry").exception(env, || 0)
    })
}

#[no_mangle]
pub extern "system" fn Java_com_surrealdb_Geometry_toString<'local>(
    mut env: EnvUnowned<'local>,
    _class: JClass<'local>,
    ptr: jlong,
) -> jstring {
    with_env_body!(env, env, {
        let value = get_value_instance!(env, ptr, null_mut);
        if let Value::Geometry(g) = value.as_ref() {
            return new_string!(env, g.to_string(), null_mut);
        }
        SurrealError::NullPointerException("Geometry").exception(env, null_mut)
    })
}
